### PR TITLE
Initial commit for introducing the SONiC Platform Development Env (Re…

### PR DESCRIPTION
**Pull request by geans-pin**
What I did
This PR is a resubmit of PR#3778

How I did it
The PDE silicon test harness and platform test harness can be found in
src/sonic-platform-pdk-pde

How to verify it
You must have both this PR and PR below.
Azure/sonic-platform-pdk-pde#28

Enable PDE by un-commenting line in rules/config
INCLUDE_PDE=y

Currently only Broadcom support has been added.

Description for the changelog
Adding support for Platform Development Environment (PDE)

### CODE REVIEW

File: Makefile.work [hunk 1]
```diff
@@ -241,6 +241,8 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            SONIC_IMAGE_VERSION=$(SONIC_IMAGE_VERSION) \
                            ENABLE_DHCP_GRAPH_SERVICE=$(ENABLE_DHCP_GRAPH_SERVICE) \
                            ENABLE_ZTP=$(ENABLE_ZTP) \
+                           ENABLE_PDE=$(ENABLE_PDE) \
+
```
Comment (xumia) [inline Makefile.work:245]: The empty line should be removed.

Comment (geans-pin) [reply Makefile.work:245]: Thanks Xumia, got it let me fix this

**Comment by adyeung** [comment]
@jleveque Pls help review

File: Makefile.work [hunk 2]
```diff
@@ -340,6 +352,11 @@ showtag:
 	@echo $(SLAVE_IMAGE):$(SLAVE_TAG)
 	@echo $(SLAVE_BASE_IMAGE):$(SLAVE_BASE_TAG)
 
+initpde:
+	$(QUIET)for submodule in $(PDESUBMODULES) ; do \
```
Comment (jleveque) [inline Makefile.work:356]: Where is `QUIET` defined?

Comment (geans-pin) [reply Makefile.work:356]: I will fix the typo.

Comment (geans-pin) [reply Makefile.work]: Can you check the fix on the latest commit ?

Geans

File: dockers/docker-pde/base_image_files/pde-bench [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+# Launch the benchmark test suite
+docker exec -it pde /usr/bin/bench.sh
+echo "$(docker exec -it pde cat ${CSV})" > ${CSV}
+
+echo ""
+echo "CSV Report: ${CSV}"
+echo "-----------------------------------------------------------------------"
+echo "$(cat ${CSV})"
+echo "-----------------------------------------------------------------------"
+
```
Comment (jleveque) [inline dockers/docker-pde/base_image_files/pde-bench:38]: Delete trailing empty line.

File: dockers/docker-pde/docker_init.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    else
+       echo "Error: Unable to locate ${SONIC_PLATFORM_WHEEL}"
+    fi
+fi
+
+start_bcm
+
+exec /usr/bin/supervisord
+
+
```
Comment (jleveque) [inline dockers/docker-pde/docker_init.sh:45]: Delete trailing empty lines.

File: dockers/docker-pde/stress-sfp-i2c.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        sys.stdout.write("#")
+        sys.stdout.flush()
+    print("\nPASS")
+
+if __name__ == '__main__':
+    sec = 180
+    if len(sys.argv) >= 2:
+        sec = int(sys.argv[1])
+    stress_sfp_i2c(sec)
+
```
Comment (jleveque) [inline dockers/docker-pde/stress-sfp-i2c.py:157]: Delete trailing empty line.

File: platform/broadcom/sonic-pde-tests.dep [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+DEP_FILES   := $(SONIC_COMMON_FILES_LIST) platform/broadcom/sonic-pde-tests.mk platform/broadcom/sonic-pde-tests.dep
+DEP_FILES   += $(SONIC_COMMON_BASE_FILES_LIST)
+SMDEP_FILES := $(addprefix $(SPATH)/,$(shell cd $(SPATH) && git ls-files))
+
+$(SONIC_PLATFORM_PDE)_CACHE_MODE  := GIT_CONTENT_SHA
+$(SONIC_PLATFORM_PDE)_DEP_FLAGS   := $(SONIC_COMMON_FLAGS_LIST)
+$(SONIC_PLATFORM_PDE)_DEP_FILES   := $(DEP_FILES)
+$(SONIC_PLATFORM_PDE)_SMDEP_FILES := $(SMDEP_FILES)
+$(SONIC_PLATFORM_PDE)_SMDEP_PATHS := $(SPATH)
+
```
Comment (jleveque) [inline platform/broadcom/sonic-pde-tests.dep:12]: Delete trailing empty line.

File: platform/broadcom/sonic-pde-tests.dep [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
```
Comment (jleveque) [inline platform/broadcom/sonic-pde-tests.dep:1]: Delete leading empty line.

**Review by jleveque — CHANGES_REQUESTED** [review]

**Comment by geans-pin** [comment]
jleveque,

I have fixed the review comment. However, the kvm test is fail in the checker which seems not related with the fix.
Can you re-run the pipeline build ?

Geans

**Review by jleveque — CHANGES_REQUESTED** [review]
Please fix LGTM alerts

**Comment by geans-pin** [comment]
> Please fix LGTM alerts

Done, with new commit

**Review by jleveque — APPROVED** [review]

**Comment by adyeung** [comment]
@lguohan @jleveque pls help resubmit test

**Comment by jleveque** [comment]
> @lguohan @jleveque pls help resubmit test

@adyeung: There is now a conflict. Please fix the conflict. That will also re-trigger the check builds.

File: dockers/docker-pde/stress-sfp-i2c.py [hunk 8]
```diff
@@ -151,4 +155,17 @@ def stress_sfp_i2c(sec=180):
     sec = 180
     if len(sys.argv) >= 2:
         sec = int(sys.argv[1])
+
+
+
+
+
+
+
```
Comment (jleveque) [inline dockers/docker-pde/stress-sfp-i2c.py:164]: Why add all these newlines?

Comment (geans-pin) [reply dockers/docker-pde/stress-sfp-i2c.py:164]: Looks like caused by removing conflicts, I can remove this

**Comment by geans-pin** [comment]
jleveque,

Please help to merged. The empty line issue is fixed.

Geans

**Comment by adyeung** [comment]
@jleveque @lguohan pls help merge

**Review by jleveque — DISMISSED** [review]
Looks good to me. Please wait for other reviewers.

**Comment by jleveque** [comment]
@qiluo-msft, @xumia: Can you please also help review this PR, as it is a bit large. Just want to get other sets of eyes on it. Thanks!

**Review by xumia — APPROVED** [review]

**Comment by qiluo-msft** [comment]
There is no `sonic-platform-pdk-pde` submodule in sonic-buildimage. Do you want to add it?

**Comment by geans-pin** [comment]
> There is no `sonic-platform-pdk-pde` submodule in sonic-buildimage. Do you want to add it?

FYI, It's already there...
https://github.com/Azure/sonic-platform-pdk-pde/pull/28

**Comment by qiluo-msft** [comment]
You did not get my question. It looks there is no relationship between sonic-buildimage repo and sonic-platform-pdk-pde repo. Will you add a submodule for sonic-platform-pdk-pde in sonic-buildimage repo?

**Comment by geans-pin** [comment]
> You did not get my question. It looks there is no relationship between sonic-buildimage repo and sonic-platform-pdk-pde repo. Will you add a submodule for sonic-platform-pdk-pde in sonic-buildimage repo?

Yes, we will need a submodule for sonic-platform-pdk-pde in sonic-buildimage.
The submodule will locate at => src/sonic-platform-pde 

What else we need to do for creating a submodule to link with sonic-platform-pdk-pde in community ?
Can we merge the PR7510 first and then create a new PR for addressing the sub module link ?

Geans

**Comment by qiluo-msft** [comment]
Better add the submodule in this PR.

**Comment by geans-pin** [comment]
> Better add the submodule in this PR.

Do we have a reference for add sub

> Better add the submodule in this PR.

Not sure how can I do for this on the community ?
Wonder if we have a reference to see how to do this ? Can you advise ?

Geans

**Comment by geans-pin** [comment]
QILUO, please check the submodule link in .gitmodule and let me know if this is ok

[submodule "src/sonic-platform-pde"]
path = src/sonic-platform-pde
url = https://github.com/Azure/sonic-platform-pdk-pde.git

Geans

**Comment by qiluo-msft** [comment]
Thanks! I think you also need `git add` the submodule's folder.
I found an example https://github.com/Azure/sonic-buildimage/pull/5700/files

**Comment by qiluo-msft** [comment]
Sorry for misleading. I notice an unexpected file src/sonic-platform-pde/.gitkeep 
Please double check the page https://git-scm.com/book/en/v2/Git-Tools-Submodules

**Comment by geans-pin** [comment]
Thanks qiluo, Here we are.. Just committed the fix

**Review by qiluo-msft — COMMENTED** [review]
The submodule looks good now. Please check with other reviewers.

**Comment by geans-pin** [comment]
xumia, jleveque, lguohan,

All of the reviewed comments are fixed and build are Pass.
Are we ready for the merge ?

Geans

**Comment by anshuv-mfst** [comment]
Hi @xumia , @qiluo-msft - can you please approve and merge asap, this PR is needed in 202106 release.

**Comment by adyeung** [comment]
@xumia, @qiluo-msft, @lguohan, @anshuv-mfst, pls help approve and merge

File: rules/config [hunk 9]
```diff
@@ -51,6 +51,8 @@ DEFAULT_PASSWORD = YourPaSsWoRd
 # ENABLE_ZTP - installs Zero Touch Provisioning support.
 # ENABLE_ZTP = y
 
+# ENABLE_PDE - Enable platform development enviroment
+ENABLE_PDE = y
```
Comment (lguohan) [inline rules/config:55]: please use INCLUDE_XXXX semantic, see example below

File: platform/broadcom/one-pde-image.mk [hunk 10]
```diff
@@ -0,0 +1,81 @@
+# sonic broadcom one image installer
+
+SONIC_ONE_PDE_IMAGE = sonic-broadcom-pde.bin
```
Comment (lguohan) [inline platform/broadcom/one-pde-image.mk:3]: it seems it is not necessary to define a entire new sonic pde image for this purpose.
we can use existing sonic image and introduce a INCLUDE_SONIC_CORE = n to disable the all the core components, then we can add INCLUDE_PDE that only include PDE docker.

This could be a cleaner design.

File: slave.mk [hunk 11]
```diff
@@ -1065,6 +1068,7 @@ $(addprefix $(TARGET_PATH)/, $(SONIC_INSTALLERS)) : $(TARGET_PATH)/% : \
 			./build_image.sh $(LOG)
 	)
 
+
```
Comment (lguohan) [inline slave.mk:1071]: please remove unnecessary lines

File: slave.mk [hunk 12]
```diff
@@ -241,6 +241,7 @@ $(info "HTTP_PROXY"                      : "$(HTTP_PROXY)")
 $(info "HTTPS_PROXY"                     : "$(HTTPS_PROXY)")
 $(info "NO_PROXY"                        : "$(NO_PROXY)")
 $(info "ENABLE_ZTP"                      : "$(ENABLE_ZTP)")
+$(info "ENABLE_PDE"                      : "$(ENABLE_PDE)")
```
Comment (lguohan) [inline slave.mk:244]: use INCLUDE_PDE

File: Makefile.work [hunk 13]
```diff
@@ -346,6 +358,14 @@ showtag:
 	@echo $(SLAVE_IMAGE):$(SLAVE_TAG)
 	@echo $(SLAVE_BASE_IMAGE):$(SLAVE_BASE_TAG)
 
+initpde:
```
Comment (lguohan) [inline Makefile.work:361]: this does not seem necessary, why not checkout all the source code.

Comment (geans-pin) [reply Makefile.work:361]: This is for generating light weight SONIC for co-working with ODM on the HWQ.
We will run make initpde to generate the SONIC PDE source package.

Comment (qiluo-msft) [reply Makefile.work:361]: The git operations should be out of the scope of Makefile. Every developer should recursively fetch the full source tree, if you want to build something in small scope, you don't need to limit the code checkout process.

File: platform/broadcom/one-pde-image.mk [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+$(SONIC_ONE_PDE_IMAGE)_MACHINE = broadcom
+$(SONIC_ONE_PDE_IMAGE)_IMAGE_TYPE = onie
+$(SONIC_ONE_PDE_IMAGE)_INSTALLS += $(BRCM_OPENNSL_KERNEL)
+$(SONIC_ONE_PDE_IMAGE)_INSTALLS += $(PDDF_PLATFORM_MODULE)
+$(SONIC_ONE_PDE_IMAGE)_LAZY_INSTALLS += $(DELL_S6000_PLATFORM_MODULE) \
```
Comment (qiluo-msft) [inline platform/broadcom/one-pde-image.mk:8]: >$(SONIC_ONE_PDE_IMAGE)_LAZY_INSTALLS [](http://example.com/codeflow?start=0&length=36)

This is duplicated as `$(SONIC_ONE_IMAGE)_LAZY_INSTALLS`, hard to maintain.

File: slave.mk (hunk 11 above)
Comment (geans-pin) [reply slave.mk:1071]: Removed

File: platform/broadcom/one-pde-image.mk (hunk 10 above)
Comment (geans-pin) [reply platform/broadcom/one-pde-image.mk]: Removed

File: platform/broadcom/one-pde-image.mk (hunk 14 above)
Comment (geans-pin) [reply platform/broadcom/one-pde-image.mk]: Removed

File: Makefile.work [hunk 15]
```diff
@@ -277,6 +278,17 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            SLAVE_DIR=$(SLAVE_DIR) \
                            $(SONIC_OVERRIDE_BUILD_VARS)
 
+PDESUBMODULES = src/sonic-linux-kernel \
```
Comment (qiluo-msft) [inline Makefile.work:281]: >PDESUBMODULES [](http://example.com/codeflow?start=0&length=13)

Also remove this list?

Comment (geans-pin) [reply Makefile.work:281]: removed

File: Makefile.work [hunk 16]
```diff
@@ -247,6 +247,7 @@ SONIC_BUILD_INSTRUCTION :=  make \
                            SONIC_IMAGE_VERSION=$(SONIC_IMAGE_VERSION) \
                            ENABLE_DHCP_GRAPH_SERVICE=$(ENABLE_DHCP_GRAPH_SERVICE) \
                            ENABLE_ZTP=$(ENABLE_ZTP) \
+                           ENABLE_PDE=$(ENABLE_PDE) \
```
Comment (lguohan) [inline Makefile.work:250]: why do we still have ENABLE_PDE? isn't that you have changed them to INCLUDE_PDE?

File: slave.mk [hunk 17]
```diff
@@ -1042,6 +1044,7 @@ $(addprefix $(TARGET_PATH)/, $(SONIC_INSTALLERS)) : $(TARGET_PATH)/% : \
 
 		DEBUG_IMG="$(INSTALL_DEBUG_TOOLS)" \
 		DEBUG_SRC_ARCHIVE_FILE="$(DBG_SRC_ARCHIVE_FILE)" \
+		BUILD_TARGET="$@" \
```
Comment (lguohan) [inline slave.mk:1047]: i am not sure why this change is needed, can you explain?

File: slave.mk (hunk 11 above)
Comment (lguohan) [inline slave.mk:1071]: please remove blank lines.

File: Makefile.work (hunk 16 above)
Comment (geans-pin) [reply Makefile.work:250]: Need to changed

File: slave.mk (hunk 17 above)
Comment (geans-pin) [reply slave.mk:1047]: Removed

File: slave.mk (hunk 11 above)
Comment (geans-pin) [reply slave.mk:1071]: removed

File: slave.mk (hunk 17 above)
Comment (geans-pin) [reply slave.mk]: See the reply from the same question from you to Bill last year, removed this for now =>

@lguohan lguohan on 29 Feb 2020 Contributor
I do not this BUILD_TARGET used anyway. is there something else missing?
 
@wbschwar wbschwar on 24 Sep 2020 Author
We pass the BUILD_TARGET to the docker build slave so we can support the build target sonic-broadcom-pde.bin to differenciate it from the normal build. Please let me know if there is a better way to handle this.

Comment (geans-pin) [reply slave.mk]: Looks like the build is fail now. Remove BUILD_TARGET="$@" \ should not affect the build

**Review by lguohan — APPROVED** [review]
:shipit:

**Comment by MikeHcChen** [comment]
Recently, I tried to build PDE environment with 202111 branch codes.
I added "INCLUDE_PDE = y" and "PDDF_SUPPORT = y" in "sonic-buildimage/rules/config" file and 
revised "sonic-buildimage/dockers/docker-pde/Dockerfile.j2" as following:
_--- a/dockers/docker-pde/Dockerfile.j2
+++ b/dockers/docker-pde/Dockerfile.j2
@@ -1,4 +1,4 @@
-FROM {{ docker_pde_load_image }}
+FROM docker-config-engine-buster_

Then building PDE using "make init;make configure PLATFORM=broadcom; **make target/sonic-broadcom-pde.bin**". I found I can get docker-pde.gz (PDE docker) and sonic-platform-pde_1.0_amd64.deb, but cannot get 
**sonic-broadcom-pde.bin**. The error logs like:

_make: *** No rule to make target 'target/sonic-broadcom-pde.bin'.  Stop.
Makefile.work:304: recipe for target 'target/sonic-broadcom-pde.bin' failed
make[1]: *** [target/sonic-broadcom-pde.bin] Error 2
make[1]: Leaving directory '/home/mike/sonic_202111_branch/sonic-buildimage'
Makefile:30: recipe for target 'target/sonic-broadcom-pde.bin' failed
make: *** [target/sonic-broadcom-pde.bin] Error 2_

Is there anything I should add to make the PDE image in SONiC?
Any idea is welcome. Thanks.
